sqlTask.sql

```diff
@@ -0,0 +1,737 @@
+WHERE Clients.StatusId = @updateStatus;
+
+EXEC AddTenToBalanceByStatusId @updateStatus;
+
+SELECT Accounts.Id AS AccountId, Accounts.Balance, Clients.Id AS ClientId, Clients.StatusId
+FROM Accounts
+	JOIN Clients ON Accounts.ClientId = Clients.Id
+WHERE Clients.StatusId = @updateStatus;
+
+GO
+
+/*
+	6. Получить список доступных средств для каждого клиента.
+	Если на аккаунте 60, а на 2х карточках по 15, тогда доступно 30
+	для перевода на карточку
+*/
+
+SELECT Accounts.Id AS AccountId, (Accounts.Balance - ISNULL(SUM(Cards.Balance), 0)) AS FreeMoney
+FROM Accounts
+	LEFT JOIN Cards ON Accounts.Id =  Cards.AccountId
+GROUP BY Accounts.Id, Accounts.Balance
+
+GO
+
+/*
+	7. Написать процедуру которая будет переводить определённую сумму
+	со счёта на карту этого аккаунта.  При этом будем считать что 
+	деньги на счёту все равно останутся, просто сумма средств на карте увеличится.
+	Переводить БЕЗОПАСНО. То есть использовать транзакцию
+*/
+
+CREATE PROCEDURE SendMoneyToCard
+	@cardId INT,
+	@moneyToSend MONEY
+AS
+BEGIN
+	DECLARE @freeMoney INT;
+
+	IF @moneyToSend <= 0
+		THROW 60004, 'Invalid parameter @moneyToSend', 1;
+
+	IF (SELECT Id FROM Cards WHERE Id = @cardId) IS NULL
+		THROW 60002, 'Card does not exist', 1;
+	
+	SET @freeMoney = (SELECT (AccountBalance - CardsBalance.CardsSumBalance) AS FreeMoney
+					  FROM (SELECT Accounts.Id AS AccountId, Accounts.Balance AS AccountBalance
+						    FROM Accounts
+						    WHERE Accounts.Id = (SELECT Cards.AccountId
+											     FROM Cards
+											     WHERE Cards.Id = @cardId)) AS Account
+					      JOIN (SELECT AccountId, SUM(Cards.Balance) AS CardsSumBalance
+							    FROM Cards
+							    GROUP BY AccountId) AS CardsBalance ON Account.AccountId = CardsBalance.AccountId)
+
+	IF @moneyToSend > @freeMoney
+		THROW 60003, 'Not enough money', 1;
+
+	BEGIN TRANSACTION
+
+	UPDATE Cards
+	SET Balance = Balance + @moneyToSend
+	WHERE Cards.Id = (SELECT Id 
+					  FROM Cards 
+					  WHERE Id = @cardId)
+
+	if (@@ERROR <> 0)
+			ROLLBACK
+
+	COMMIT
+
+END;
+
+GO
+
+/*
+	Тест процедуры SendMoneyToCard. Ошибок не ожидается. 
+	Лимит для перевода на заданноую карту 68
+*/
+
+DECLARE @updateCard INT, @moneyToSend INT;
+SET @moneyToSend = 23
+SET @updateCard = (SELECT Cards.Id 
+				   FROM Cards 
+				   WHERE Cards.CardNumber = '1673014554636870')
+
+SELECT Accounts.Id AS AccountId, Accounts.Balance AS AccountBalance, Cards.Id AS CardId, Cards.Balance AS CardBalance
+FROM Cards
+	JOIN Accounts ON Accounts.Id = Cards.AccountId
+WHERE Cards.Id =  @updateCard;
+
+EXEC SendMoneyToCard @updateCard, @moneyToSend;
+
+SELECT Accounts.Id AS AccountId, Accounts.Balance AS AccountBalance, Cards.Id AS CardId, Cards.Balance AS CardBalance
+FROM Cards
+	JOIN Accounts ON Accounts.Id = Cards.AccountId
+WHERE Cards.Id =  @updateCard;
+
+GO
+
+/*
+	8. Написать триггер на таблицу Account
+	нельзя изменить значение в Account на меньшее, 
+	чем сумма балансов по всем карточкам
```

Comment:
8.1) Все-таки хотелось бы получать ошибку, если получается, что средств на счете станет меньше - игнорировать неправильное поведение не лучшая идея. Будем считать, что весь Update некорректный, если хотя бы в одном счете станет меньше средств, чем на карточках.
Давай сделаем запрос попроще и табличная переменная нам тоже не понадобится.

Reply:
Если есть желание, то можешь переделать триггер на After вместо Instead Of, тогда просто в случае чего сделаешь ROLLBACK TRAN, не надо будет писать самому Update в триггере